src/ucs/datastruct/ptr_array.h

```diff
@@ -120,13 +150,229 @@ void *ucs_ptr_array_replace(ucs_ptr_array_t *ptr_array, unsigned index, void *ne
+typedef struct ucs_ptr_array_locked {
+    ucs_ptr_array_t          super;
+    ucs_recursive_spinlock_t lock;
+} ucs_ptr_array_locked_t;
+
+
+/**
+ * Locked array init
+ *
+ * @param [in] locked_ptr_array  Pointer to a locked ptr array.
+ * @param [in] init_placeholder  Default placeholder value.
+ * @param [in] name              The name of the ptr array.
+ *
+ * @return Success or failure.
+ *
+ */
+ucs_status_t
+ucs_ptr_array_locked_init(ucs_ptr_array_locked_t *locked_ptr_array,
+                          uint32_t init_placeholder,
+                          const char *name);
+
+
+/**
+ * Cleanup the locked array.
+ *
+ * @param [in] locked_ptr_array    Pointer to a locked ptr array.
+ *
+ * @note All values should already be removed from it.
+ */
+void ucs_ptr_array_locked_cleanup(ucs_ptr_array_locked_t *locked_ptr_array);
+
+
+/**
+ * Insert a pointer to the locked array.
+ *
+ * @param [in]  locked_ptr_array  Pointer to a locked ptr array.
+ * @param [in]  value             Pointer to insert. Must be 8-byte aligned.
+ * @param [out] placeholder       Filled with placeholder value.
+ *
+ * @return The index to which the value was inserted.
+ *
+ * Complexity: Amortized O(1)
+ *
+ * @note The array will grow if needed.
+ */
+unsigned ucs_ptr_array_locked_insert(ucs_ptr_array_locked_t *locked_ptr_array,
+                                     void *value, uint32_t *placeholder_p);
+
+
+/**
+ * Remove a pointer from the locked array.
+ *
+ * @param [in] locked_ptr_array  Pointer to a locked ptr array.
+ * @param [in] index             Index to remove from.
+ * @param [in] placeholder       Value to put in the free slot.
+ *
+ * Complexity: O(1)
+ */
+void ucs_ptr_array_locked_remove(ucs_ptr_array_locked_t *locked_ptr_array,
+                                 unsigned index, uint32_t placeholder);
+
+
+/**
+ * Replace pointer in the locked array
+ *
+ * @param [in] locked_ptr_array  Pointer to a locked ptr array.
+ * @param [in] index             Index of slot
+ * @param [in] new_val           Value to put into slot given by index
+ *
+ * @return Old value of the slot
+ *
+ * Complexity: O(1)
+ */
+void *ucs_ptr_array_locked_replace(ucs_ptr_array_locked_t *locked_ptr_array,
+	                           unsigned index, void *new_val);
+
+
+/** Acquire the ptr_array lock
+ *
+ * @param [in] locked_ptr_array  Pointer to a locked ptr array.
+ */
+#define ucs_ptr_array_locked_acquire_lock(_locked_ptr_array) \
+    ucs_recursive_spin_lock(&(_locked_ptr_array)->lock)
+
+
+/** Release the ptr_array lock
+ *
+ * @param [in] locked_ptr_array  Pointer to a locked ptr array.
+ */
+#define ucs_ptr_array_locked_release_lock(_locked_ptr_array) \
+    ucs_recursive_spin_unlock(&(_locked_ptr_array)->lock)
+
+
+/**
+ * Retrieves a value from the locked array.
+ *
+ * The user is required to lock using ucs_ptr_array_locked_acquire_lock()
+ * and ucs_ptr_array_locked_release_lock() functions.
+ *
+ * @param [in]  _locked_ptr_array   Pointer to a locked ptr array.
```

Comment:
Since this is function, no need for `_` prefix
Also, check other places

Reply:
Fixed.